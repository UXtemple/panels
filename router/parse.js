import withTrailingSlash from './with-trailing-slash'

const DEFAULT_PARSER = /^https?:\/\/([a-zA-Z0-9\-_.]+)()(\/.*)/
const PROTOCOL = /(https?):\/\//
const SLICE_END = ')'
const SLICE_START = '('
const SLICE_MARKERS = new RegExp(`[${SLICE_START}${SLICE_END}]`, 'g')

// TODO review this, instead of discarding the query string perhaps we can pass it down to
// the last panel?
const withoutQueryString = s => (/\?/.test(s) ? s.match(/(.*?)\?/)[1] : s)

const MANY_APPS = /^(https?:\/\/.+?\/)(https?:\/\/.+)/

const appSplit = s => {
  let currentUri = s
  let nextUri = false

  const manyAppsMatch = s.match(MANY_APPS)
  if (manyAppsMatch) {
    currentUri = manyAppsMatch[1]
    nextUri = manyAppsMatch[2]
  }

  return {
    currentUri,
    nextUri,
  }
}

export default function parse(uri, parsers = []) {
  const apps = {
    byName: {},
    items: [],
  }
  const routes = {
    byContext: {},
    items: [],
  }

  // Make sure we always have a trailing slash on the URI
  const protocol = uri.match(PROTOCOL)[1]
  let nextUri = withoutQueryString(withTrailingSlash(uri))

  do {
    const split = appSplit(nextUri)
    const currentUri = split.currentUri
    nextUri = split.nextUri

    const parser = parsers.find(p => p.test(currentUri)) || DEFAULT_PARSER
    let [_0, app, _1, path] = currentUri.match(parser)

    const base = `${protocol}://${app}`
    const context = routes.items.length > 0
      ? routes.items[routes.items.length - 1]
      : ''

    // Get every path 'bit' which is indeed every panel we need to load
    let pathRoute = []
    let isVisible = true
    do {
      path = path.split('/')
      const lastBit = path.length > 1 ? path[path.length - 2] : ''
      path = path.slice(0, path.length - 1).join('/')
      const hasSliceEndMarkerForRoot = lastBit.indexOf(SLICE_END) === 0
      const hasSliceEndMarker =
        !hasSliceEndMarkerForRoot && lastBit.indexOf(SLICE_END) > -1
      const hasSliceStartMarker = lastBit.indexOf(SLICE_START) > -1

      if (hasSliceEndMarker || hasSliceStartMarker) {
        isVisible = false
      }

      const panelPath = path.replace(SLICE_MARKERS, '') || '/'

      const panelId = `${app}${panelPath}`

      pathRoute.push({
        app,
        context: `${context}${base}${withTrailingSlash(path)}`,
        isVisible,
        panelId,
        path: panelPath,
      })

      isVisible = hasSliceEndMarkerForRoot
        ? false
        : hasSliceStartMarker ? true : isVisible
    } while (path.length)

    if (apps.items.indexOf(app) === -1) {
      apps.items.push(app)
      apps.byName[app] = {
        app,
        panels: [],
      }
    }

    pathRoute.reverse().forEach(route => {
      routes.byContext[route.context] = route
      routes.items.push(route.context)

      if (apps.byName[app].panels.indexOf(route.path) === -1) {
        apps.byName[app].panels.push(route.path)
      }
    })
  } while (nextUri)

  return {
    apps,
    routes,
  }
}
