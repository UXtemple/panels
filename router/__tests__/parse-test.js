import normaliseUri from 'panels-normalise-uri';
import parse from '../parse';
import test from 'tape';
import util from 'util';

const cases = [{
  uri: 'https://UXtemple.com/',
  name: 'basic: one panel https://UXtemple.com/'
},{
  uri: 'https://UXtemple.com/panels',
  name: 'basic: many panels https://UXtemple.com/panels'
}, {
  uri: 'https://UXtemple.com/panels/../back',
  name: 'basic: backwards panels https://UXtemple.com/panels/../back'
}, {
  uri: 'https://UXtemple.com/https://usepanels.com/',
  name: 'teleport: basic https://UXtemple.com/https://usepanels.com/'
}, {
  uri: 'https://UXtemple.com/)panels',
  name: 'teleport: slice the root https://UXtemple.com/)panels'
}, {
  uri: 'https://UXtemple.com/(panels)',
  name: 'teleport: slice all panels https://UXtemple.com/(panels) i.e., hidden gems :)?'
}, {
  uri: 'https://UXtemple.com/(panels/use)',
  name: 'teleport: slice multiple panels https://UXtemple.com/(panels/use)'
}, {
  uri: 'https://UXtemple.com/(panels)/use',
  name: 'teleport: slice in between https://UXtemple.com/(panels)/use'
}, {
  uri: 'https://UXtemple.com/(panels)/use/(new)',
  name: 'teleport: slice some parts https://UXtemple.com/(panels)/use/(new)'
}, {
  uri: 'https://UXtemple.com/panels)/use/https://usepanels.com/UXtemple',
  name: 'teleport: slice one app only https://UXtemple.com/panels)/use/https://usepanels.com/UXtemple'
},
{
  uri: 'https://UXtemple.com/panels)/use/https://usepanels.com/)UXtemple',
  name: 'teleport: slice many apps https://UXtemple.com/panels)/use/https://usepanels.com/)UXtemple'
}, {
  uri: 'https://UXtemple.com/panels/https://usepanels.com/..',
  name: 'teleport: backwards panels https://UXtemple.com/panels/https://usepanels.com/..'
}]


test('#parse', t => {
  cases.forEach(c => {
    console.log(`${c.name} \n ${util.inspect(parse(c.uri), false, null)}\n`);
  });

  t.end();
});
