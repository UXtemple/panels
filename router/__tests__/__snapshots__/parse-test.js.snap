exports[`test basic: backwards panels https://UXtemple.com/panels/../back 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/panels",
          "/panels/..",
          "/panels/../back",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/panels/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/panels/",
        "isVisible": true,
        "panelId": "UXtemple.com/panels",
        "path": "/panels",
      },
      "https://UXtemple.com/panels/../": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/panels/../",
        "isVisible": true,
        "panelId": "UXtemple.com/panels/..",
        "path": "/panels/..",
      },
      "https://UXtemple.com/panels/../back/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/panels/../back/",
        "isVisible": true,
        "panelId": "UXtemple.com/panels/../back",
        "path": "/panels/../back",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/panels/",
      "https://UXtemple.com/panels/../",
      "https://UXtemple.com/panels/../back/",
    ],
  },
}
`;

exports[`test basic: many panels https://UXtemple.com/panels 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/panels",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/panels/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/panels/",
        "isVisible": true,
        "panelId": "UXtemple.com/panels",
        "path": "/panels",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/panels/",
    ],
  },
}
`;

exports[`test basic: one panel https://UXtemple.com/ 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
    ],
  },
}
`;

exports[`test routerWhitelist: one specific domain and a standard domain https://specificDomain.com/root/https://UXtemple.com/ 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
        ],
      },
      "specificDomain.com/root": Object {
        "app": "specificDomain.com/root",
        "panels": Array [
          "/",
        ],
      },
    },
    "items": Array [
      "specificDomain.com/root",
      "UXtemple.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://specificDomain.com/root/": Object {
        "app": "specificDomain.com/root",
        "context": "https://specificDomain.com/root/",
        "isVisible": true,
        "panelId": "specificDomain.com/root/",
        "path": "/",
      },
      "https://specificDomain.com/root/https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://specificDomain.com/root/https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
    },
    "items": Array [
      "https://specificDomain.com/root/",
      "https://specificDomain.com/root/https://UXtemple.com/",
    ],
  },
}
`;

exports[`test routerWhitelist: one specific domain and panel and a standard domain with panel https://specificDomain.com/root/panelhttps://UXtemple.com/otherPanel 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/otherPanel",
        ],
      },
      "specificDomain.com/root": Object {
        "app": "specificDomain.com/root",
        "panels": Array [
          "/",
        ],
      },
    },
    "items": Array [
      "specificDomain.com/root",
      "UXtemple.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://specificDomain.com/root/": Object {
        "app": "specificDomain.com/root",
        "context": "https://specificDomain.com/root/",
        "isVisible": true,
        "panelId": "specificDomain.com/root/",
        "path": "/",
      },
      "https://specificDomain.com/root/https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://specificDomain.com/root/https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://specificDomain.com/root/https://UXtemple.com/otherPanel/": Object {
        "app": "UXtemple.com",
        "context": "https://specificDomain.com/root/https://UXtemple.com/otherPanel/",
        "isVisible": true,
        "panelId": "UXtemple.com/otherPanel",
        "path": "/otherPanel",
      },
    },
    "items": Array [
      "https://specificDomain.com/root/",
      "https://specificDomain.com/root/https://UXtemple.com/",
      "https://specificDomain.com/root/https://UXtemple.com/otherPanel/",
    ],
  },
}
`;

exports[`test teleport: backwards panels https://UXtemple.com/panels/https://usepanels.com/.. 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/panels",
        ],
      },
      "usepanels.com": Object {
        "app": "usepanels.com",
        "panels": Array [
          "/",
          "/..",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
      "usepanels.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/panels/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/panels/",
        "isVisible": true,
        "panelId": "UXtemple.com/panels",
        "path": "/panels",
      },
      "https://UXtemple.com/panels/https://usepanels.com/": Object {
        "app": "usepanels.com",
        "context": "https://UXtemple.com/panels/https://usepanels.com/",
        "isVisible": true,
        "panelId": "usepanels.com/",
        "path": "/",
      },
      "https://UXtemple.com/panels/https://usepanels.com/../": Object {
        "app": "usepanels.com",
        "context": "https://UXtemple.com/panels/https://usepanels.com/../",
        "isVisible": true,
        "panelId": "usepanels.com/..",
        "path": "/..",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/panels/",
      "https://UXtemple.com/panels/https://usepanels.com/",
      "https://UXtemple.com/panels/https://usepanels.com/../",
    ],
  },
}
`;

exports[`test teleport: basic https://UXtemple.com/https://usepanels.com/ 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
        ],
      },
      "usepanels.com": Object {
        "app": "usepanels.com",
        "panels": Array [
          "/",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
      "usepanels.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/https://usepanels.com/": Object {
        "app": "usepanels.com",
        "context": "https://UXtemple.com/https://usepanels.com/",
        "isVisible": true,
        "panelId": "usepanels.com/",
        "path": "/",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/https://usepanels.com/",
    ],
  },
}
`;

exports[`test teleport: slice all panels https://UXtemple.com/(panels) i.e., hidden gems :)? 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/panels",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/(panels)/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/(panels)/",
        "isVisible": false,
        "panelId": "UXtemple.com/panels",
        "path": "/panels",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/(panels)/",
    ],
  },
}
`;

exports[`test teleport: slice in between https://UXtemple.com/(panels)/use 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/panels",
          "/panels/use",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/(panels)/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/(panels)/",
        "isVisible": false,
        "panelId": "UXtemple.com/panels",
        "path": "/panels",
      },
      "https://UXtemple.com/(panels)/use/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/(panels)/use/",
        "isVisible": true,
        "panelId": "UXtemple.com/panels/use",
        "path": "/panels/use",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/(panels)/",
      "https://UXtemple.com/(panels)/use/",
    ],
  },
}
`;

exports[`test teleport: slice many apps https://UXtemple.com/panels)/use/https://usepanels.com/)UXtemple 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/panels",
          "/panels/use",
        ],
      },
      "usepanels.com": Object {
        "app": "usepanels.com",
        "panels": Array [
          "/",
          "/UXtemple",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
      "usepanels.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": false,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/panels)/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/panels)/",
        "isVisible": false,
        "panelId": "UXtemple.com/panels",
        "path": "/panels",
      },
      "https://UXtemple.com/panels)/use/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/panels)/use/",
        "isVisible": true,
        "panelId": "UXtemple.com/panels/use",
        "path": "/panels/use",
      },
      "https://UXtemple.com/panels)/use/https://usepanels.com/": Object {
        "app": "usepanels.com",
        "context": "https://UXtemple.com/panels)/use/https://usepanels.com/",
        "isVisible": false,
        "panelId": "usepanels.com/",
        "path": "/",
      },
      "https://UXtemple.com/panels)/use/https://usepanels.com/)UXtemple/": Object {
        "app": "usepanels.com",
        "context": "https://UXtemple.com/panels)/use/https://usepanels.com/)UXtemple/",
        "isVisible": true,
        "panelId": "usepanels.com/UXtemple",
        "path": "/UXtemple",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/panels)/",
      "https://UXtemple.com/panels)/use/",
      "https://UXtemple.com/panels)/use/https://usepanels.com/",
      "https://UXtemple.com/panels)/use/https://usepanels.com/)UXtemple/",
    ],
  },
}
`;

exports[`test teleport: slice multiple panels https://UXtemple.com/(panels/use) 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/panels",
          "/panels/use",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/(panels/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/(panels/",
        "isVisible": false,
        "panelId": "UXtemple.com/panels",
        "path": "/panels",
      },
      "https://UXtemple.com/(panels/use)/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/(panels/use)/",
        "isVisible": false,
        "panelId": "UXtemple.com/panels/use",
        "path": "/panels/use",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/(panels/",
      "https://UXtemple.com/(panels/use)/",
    ],
  },
}
`;

exports[`test teleport: slice one app only https://UXtemple.com/panels)/use/https://usepanels.com/UXtemple 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/panels",
          "/panels/use",
        ],
      },
      "usepanels.com": Object {
        "app": "usepanels.com",
        "panels": Array [
          "/",
          "/UXtemple",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
      "usepanels.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": false,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/panels)/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/panels)/",
        "isVisible": false,
        "panelId": "UXtemple.com/panels",
        "path": "/panels",
      },
      "https://UXtemple.com/panels)/use/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/panels)/use/",
        "isVisible": true,
        "panelId": "UXtemple.com/panels/use",
        "path": "/panels/use",
      },
      "https://UXtemple.com/panels)/use/https://usepanels.com/": Object {
        "app": "usepanels.com",
        "context": "https://UXtemple.com/panels)/use/https://usepanels.com/",
        "isVisible": true,
        "panelId": "usepanels.com/",
        "path": "/",
      },
      "https://UXtemple.com/panels)/use/https://usepanels.com/UXtemple/": Object {
        "app": "usepanels.com",
        "context": "https://UXtemple.com/panels)/use/https://usepanels.com/UXtemple/",
        "isVisible": true,
        "panelId": "usepanels.com/UXtemple",
        "path": "/UXtemple",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/panels)/",
      "https://UXtemple.com/panels)/use/",
      "https://UXtemple.com/panels)/use/https://usepanels.com/",
      "https://UXtemple.com/panels)/use/https://usepanels.com/UXtemple/",
    ],
  },
}
`;

exports[`test teleport: slice some parts https://UXtemple.com/(panels)/use/(new) 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/panels",
          "/panels/use",
          "/panels/use/new",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": true,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/(panels)/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/(panels)/",
        "isVisible": false,
        "panelId": "UXtemple.com/panels",
        "path": "/panels",
      },
      "https://UXtemple.com/(panels)/use/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/(panels)/use/",
        "isVisible": true,
        "panelId": "UXtemple.com/panels/use",
        "path": "/panels/use",
      },
      "https://UXtemple.com/(panels)/use/(new)/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/(panels)/use/(new)/",
        "isVisible": false,
        "panelId": "UXtemple.com/panels/use/new",
        "path": "/panels/use/new",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/(panels)/",
      "https://UXtemple.com/(panels)/use/",
      "https://UXtemple.com/(panels)/use/(new)/",
    ],
  },
}
`;

exports[`test teleport: slice the root https://UXtemple.com/)panels 1`] = `
Object {
  "apps": Object {
    "byName": Object {
      "UXtemple.com": Object {
        "app": "UXtemple.com",
        "panels": Array [
          "/",
          "/panels",
        ],
      },
    },
    "items": Array [
      "UXtemple.com",
    ],
  },
  "routes": Object {
    "byContext": Object {
      "https://UXtemple.com/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/",
        "isVisible": false,
        "panelId": "UXtemple.com/",
        "path": "/",
      },
      "https://UXtemple.com/)panels/": Object {
        "app": "UXtemple.com",
        "context": "https://UXtemple.com/)panels/",
        "isVisible": true,
        "panelId": "UXtemple.com/panels",
        "path": "/panels",
      },
    },
    "items": Array [
      "https://UXtemple.com/",
      "https://UXtemple.com/)panels/",
    ],
  },
}
`;
